### PR TITLE
feat: added new column in form for status

### PR DESCRIPTION
**Pull request by Rithingithub**
Lineman can change the status of issue

### CODE REVIEW

File: models/officials.go [hunk 1]
```diff
@@ -23,7 +25,7 @@ type Officials struct {
 	Id         uint   `gorm:"primary_key;auto_increment" json:"id"`
 	AuthId     string `gorm:"unique" json:"auth_id"`
 	Name       string `gorm:"size:255;not null" json:"name"`
-	Role       ROLE   `sql:"type:ENUM('SE', 'AE', 'OV')" gorm:"column:role"`
+	Role       ROLE   `sql:"type:ENUM('SE', 'AE', 'OV', 'LineMan')" gorm:"column:role"`
```
Comment (sreehari2003) [inline models/officials.go:28]: make it LM instead of line man

File: models/officials.go [hunk 2]
```diff
@@ -15,6 +15,8 @@ const (
 	SE ROLE = "SE"
 	//OV - overseer
 	OV ROLE = "OV"
+	// LineMan - lineman
+	LineMan ROLE = "LineMan"
```
Comment (sreehari2003) [inline models/officials.go:19]: LM

File: router/router.go [hunk 3]
```diff
@@ -67,6 +67,11 @@ func CreateRoute(h controller.Handler) *gin.Engine {
 	v1.GET("/form", verifySession(nil), h.GetAllForm)
 	v1.GET("/form/:id", verifySession(nil), h.GetFormByID)
 
+	// accesing controller by method
+	// isDoneController := controller.NewIsDoneController(db)
+	// v1.POST("/isdone/:id/status", isDoneController.ChangeStatusHandler)
+	v1.POST("/isdone/:id/status", verifySession(nil), h.ChangeStatusHandler)
```
Comment (sreehari2003) [inline router/router.go:73]: route should be patch or put req

/form/status is enough

File: models/isdone.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+	gorm.Model
+	Id        uint        `gorm:"primary_key;auto_increment" json:"id"`
+	Approved  bool        `gorm:"default:false" json:"approved"`
+	Working   bool        `gorm:"default:false" json:"working"`
+	Completed bool        `gorm:"default:false" json:"completed"`
+	Officials []Officials `gorm:"foreignkey:OfficialID"`
+	IssueID   uint        `gorm:"not null" json:"issue_id"`
+	Issue     Issue       `gorm:"foreignKey:IssueID"`
+}
+
```
Comment (sreehari2003) [inline models/isdone.go:17]: no need of a new table

File: controller/isdone.controller.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		return
+	}
+
+	// Check if the official is authorized to change the status
+	lineMan, ok := official.(*models.Officials)
+	if !ok || lineMan.Role != models.LineMan {
+		ctx.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
+		return
+	}
+
```
Comment (sreehari2003) [inline controller/isdone.controller.go:40]: remove this file we dont want to over complicte the stuff

**Review by sreehari2003 — CHANGES_REQUESTED** [review]
update the pull request

**Comment by sreehari2003** [comment]
> Owner
> ###

@VISHNURAJESHP  leave a review of the code

File: router/router.go (hunk 3 above)
Comment (Rithingithub) [reply router/router.go:73]: Is there a need of issue id for status confirmation?

File: models/form.go [hunk 6]
```diff
@@ -21,8 +21,14 @@ type Form struct {
...
 	Nooflabours    int    `gorm:"not null" json:"nooflabours"`
-	IssueID        uint   `gorm:"uniqueIndex"`
-	OfficialID     uint   `gorm:"not null" json:"official_id"`
+	// issue id is the foreign key
+	IssueID uint  `gorm:"not null" json:"issue_id"`
+	Issue   Issue `gorm:"foreignKey:IssueID"`
+
+	OfficialID uint `gorm:"not null" json:"official_id"`
+	Approved   bool `gorm:"default:false" json:"approved"`
+	Working    bool `gorm:"default:false" json:"working"`
```
Comment (sreehari2003) [inline models/form.go:30]: Not like this it should be enum not multiple column check the role column in official for example

File: controller/form.controller.go [hunk 7]
```diff
@@ -59,64 +59,117 @@ func (h Handler) CreateForm(c *gin.Context) {
...
+		"status":   http.StatusOK,
 		"response": "Data read successfully",
 		"ok":       true,
-		"data":     Form,
+		"data":     form,
 	})
 }
+
+func (h Handler) ChangeStatusHandler(c *gin.Context) {
+	// Get the issue ID from the request URL
```
Comment (sreehari2003) [inline controller/form.controller.go:121]: Should be query

File: controller/form.controller.go [hunk 8]
```diff
@@ -59,64 +59,117 @@ func (h Handler) CreateForm(c *gin.Context) {
...
+		if err == gorm.ErrRecordNotFound {
+			c.JSON(http.StatusNotFound, gin.H{"error": "Issue entry not found"})
+			return
+		}
+		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch Issue entry"})
+		return
+	}
+
+	// Get the authenticated official (LineMan)
+	official, exists := c.Get("official")
```
Comment (sreehari2003) [inline controller/form.controller.go:136]: How does this works?

File: controller/form.controller.go [hunk 9]
```diff
@@ -59,64 +59,117 @@ func (h Handler) CreateForm(c *gin.Context) {
...
+	if !exists {
+		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to get authenticated official"})
+		return
+	}
+
+	// Check if the official is authorized to change the status
+	lineman, ok := official.(*models.Officials)
+	if !ok || lineman.Role != models.LM {
```
Comment (sreehari2003) [inline controller/form.controller.go:144]: Remove this check we have already wrote a middleware for this

File: controller/form.controller.go [hunk 10]
```diff
@@ -59,64 +59,117 @@ func (h Handler) CreateForm(c *gin.Context) {
...
+		c.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
+		return
+	}
+
+	// Update the status fields based on the request body
+	var updateData struct {
+		Working   bool `json:"working"`
```
Comment (sreehari2003) [inline controller/form.controller.go:151]: The total logic should be. Changed after changing the table

**Review by sreehari2003 — CHANGES_REQUESTED** [review]
Update the pr

**Pull request merged**